### ISSUE TITLE
torch.masked_select outputs a too large tensor filled with trash when the mask is obtained from PIL

### ISSUE THREAD

**Issue by philip-bl** [pytorch/pytorch#87206]
### 🐛 Describe the bug

In the following example, `torch.masked_select` produces a too large tensor, the first part of which is filled with the correct values while the rest is filled with some trash.

```
import torch

import PIL.Image as Image
import numpy as np

x = torch.randn(1, 4, 2)
y = torch.tensor(np.asarray(Image.new("1", (2, 4), color=1))).unsqueeze(0)
print(type(y), y.shape, y.dtype, y.is_contiguous())  # <class 'torch.Tensor'> torch.Size([1, 4, 2]) torch.bool True

result = torch.masked_select(x, y)
print(result.shape)  # torch.Size([2040])
# The line above must output 8, but instead it outputs 2040.
```

Here is a google colab that reproduces the bug: https://colab.research.google.com/drive/1C2uYFPVGZ7776Fa8_PaNu1hckgjnYkPN?usp=sharing

### Versions

PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.6
Libc version: glibc-2.26

Python version: 3.7.15 (default, Oct 12 2022, 19:14:55)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.133+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: False
CUDA runtime version: 11.2.152
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.12.1+cu113
[pip3] torchaudio==0.12.1+cu113
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.13.1
[pip3] torchvision==0.13.1+cu113
[conda] Could not collect

**Comment by ngimel** [comment]
Can't repro on master

**Comment by philip-bl** [comment]
Btw, if anyone's having the same issue and wondering, a workaround I found is to do `y = y.float().bool()`.

**Comment by ngimel** [comment]
cc @peterbell10 this is probably fixed by one of your PRs for nonstandard bool representation

**Comment by peterbell10** [comment]
I agree.  The output shape is off by a factor of 255 since 255 * 8 = 2040, and PIL uses 255 to represent True. So this is the same issue as #54789.
